test/cpp/tensorexpr/tutorial.cpp

```diff
@@ -0,0 +1,128 @@
+// TODO: Add an intro
+#include <iostream>
+#include <string>
+
+#include <torch/csrc/jit/tensorexpr/buffer.h>
+#include <torch/csrc/jit/tensorexpr/expr.h>
+#include <torch/csrc/jit/tensorexpr/function.h>
+#include <torch/csrc/jit/tensorexpr/ir.h>
+#include <torch/csrc/jit/tensorexpr/ir_printer.h>
+#include <torch/csrc/jit/tensorexpr/stmt.h>
+#include <torch/csrc/jit/tensorexpr/tensor.h>
+
+using namespace torch::jit::tensorexpr;
+
+int main(int argc, char* argv[]) {
+  // TODO: Describe what KernelScope is
+  KernelScope kernel_scope;
+  {
+    // *** Structure of tensor expressions ***
+    //
+    // A tensor expression is a tree of expressions. Each expression has a type,
+    // and that type defines what sub-expressions it the current expression has.
+    // For instance, an expression of type 'Mul' would have a type 'kMul' and
+    // two subexpressions: LHS and RHS. Each of these two sub-expressions could
+    // also be a 'Mul' or some other expression.
+    //
+    // Let's construct a simple TE:
+    Expr* lhs = new IntImm(5);
+    Expr* rhs = new Var("x", kInt);
+    Expr* mul = new Mul(lhs, rhs);
+    std::cout << "Tensor expression: " << *mul << std::endl;
+    // Prints: Tensor expression: 5 * x
+
+    // Here we created an expression representing a 5*x computation, where x is
+    // an int variable.
+
+    // Another, probably a more convenient, way to construct tensor expressions
+    // is to use so called expression handles (as opposed to raw expressions
+    // like we did in the previous example). Expression handles overload common
+    // operations and allow us to express the same semantics in a more natural
+    // way:
+    ExprHandle l = 1;
+    ExprHandle r = Var::make("x", kInt);
+    ExprHandle m = l * r;
+    std::cout << "Tensor expression: " << *m.node() << std::endl;
+    // Prints: Tensor expression: 1 * x
+
+    // In a similar fashion we could construct arbitrarily complex expressions
+    // using mathematical and logical operations, casts between various data
+    // types, and a bunch of intrinsics.
+    ExprHandle a = Var::make("a", kInt);
+    ExprHandle b = Var::make("b", kFloat);
+    ExprHandle c = Var::make("c", kFloat);
+    ExprHandle x = ExprHandle(5) * a + b / (sigmoid(c) - 3.0f);
+    std::cout << "Tensor expression: " << *x.node() << std::endl;
+    // Prints: Tensor expression: float(5 * a) + b / ((sigmoid(c)) - 3.f)
+
+    // An ultimate purpose of tensor expressions is to optimize tensor
+    // computations, and in order to represent accesses to tensors data, there
+    // is a special kind of expression - a load.
+    // To construct a load we need two pieces: the base and the indices. The
+    // base of a load is a Buf expression, which could be thought of as a
+    // placeholder similar to Var, but with dimensions info.
+    //
+    // Let's construct a simple load:
+    BufHandle A("A", {ExprHandle(64), ExprHandle(32)}, kInt);
+    Placeholder A_buf(A);
```

Comment:
Is there a quick way to explain why a Placeholder is needed to wrap the BufHandle?

Reply:
The primary reason is to expose a separate class in users' API while hiding internal classes (the definition of users is vague here, but you could think of several layers of details in API: the highest level of abstraction is when you use `Compute` API and never construct expressions manually). In this sense, `Tensor` and `Placeholder` are what we expose and `Buf` is what they are built upon internally.

But technically `Placeholder` doesn't add anything to what `Buf` already has.